File: stanford algorithm class/inversion.py
#!/usr/bin/python
# see inversion.cpp for problem description
# Author: Pan Gu
def count(l):
    n = len(l);
    if n < 2: return 0, l;
    m = n/2;
    nl, ll = count(l[:m]);
    nr, lr = count(l[m:]);
    ret = nl + nr;
    i, j = 0, 0;
    for x in ll:
	while i < n-m and lr[i] < x:
	    l[j] = lr[i];
	    i, j = i+1, j+1;
	ret += i;
	l[j] = x;
	j += 1;
    l[j:] = lr[i:];
    return ret, l;
f = open("IntegerArray.txt");
l = [int(f.readline()) for i in xrange(100000)];
print count(l)[0];

File: stanford algorithm class/inversion.cpp
#include <cstdio>
#define N 100000
typedef long long ll;
int a[N+10], buf[N+10];
ll count(int *a, int n)
{
	if(n < 2) return 0;
	int m = n/2;
	ll ret = count(a, m) + count(a+m, n-m);
	int i, j, k;
	for(i = 0, j = m, k = 0; i < m; i++) {
		while(j < n && a[j] < a[i])
			buf[k++] = a[j++];
		ret += j - m;
		buf[k++] = a[i];
	}
	for(i = 0; i < k; i++) a[i] = buf[i];
	return ret;
}
int main()
{
	freopen("IntegerArray.txt", "r", stdin);
	for(int i = 0; i < N; i++) scanf("%d", &a[i]);
	printf("%lld\n", count(a, N));
}
